File: 프로그래머스/lv1/김원래/삼총사.cpp
#include <string>
#include <vector>

using namespace std;

int visit[14];
int res;


void backtrack(int studentCnt, int curSum, int curIdx, vector<int>& number)
{
    if(studentCnt == 3) {
        if(curSum == 0) ++res;
        return;
    }
    
    for(int cnt = curIdx ; cnt < number.size() ; ++cnt) {
        if(visit[cnt] == 1) continue;
        visit[cnt] = 1;
        backtrack(studentCnt + 1 , curSum + number[cnt], cnt,number);
        visit[cnt] = 0;
    }
}

int solution(vector<int> number) {
    int answer = 0;
    
    backtrack(0, 0, 0,number);
    
    answer = res;
    
    return answer;
}

File: 프로그래머스/lv1/이선영/삼총사.java
class Solution {
    static int answer = 0;
    public int solution(int[] number) {
        
        boolean[] selected = new boolean[number.length];
        solve(number,0, 0, 0, new boolean[number.length]);
        
        return answer;
    }
    
    public void solve(int[] number, int start, int sum, int cnt, boolean[] selected){
        if(cnt == 3){
            if(sum==0) answer++;
            return;
        }
        
        for(int i=start; i<number.length; i++){
                selected[i] = true;
                solve(number, i+1, sum+number[i], cnt+1, selected);
                selected[i] = false;
            
        }
    }
}
